File: Swift/剑指 Offer II/114.外星文字典.swift
//
//  114.外星文字典.swift
//  swift_
//
//  Created by kingsoft on 2022/5/31.
//

// @lc code=start
class Solution {
    func alienOrder(_ words: [String]) -> String {
        // 邻接矩阵
        var g: [Character: Set<Character>] = [:]
        // 存每个字母的入度
        var d: [Character: Int] = [:]
        // 初始化每个字母的入度
        for word in words {
            for c in word {
                d[c] = 0
            }
        }
        
        for i in 1..<words.count {
            let word1 = Array(words[i - 1])
            let word2 = Array(words[i])
            let len = min(word1.count, word2.count)
            var j = 0
            while j < len {
                if word1[j] == word2[j] {
                    j += 1
                    continue
                }
                if !(g[word1[j]]?.contains(word2[j]) ?? false) {
                    if g[word1[j]] != nil {
                        g[word1[j]]?.insert(word2[j])
                    } else {
                        g[word1[j]] = [word2[j]]
                    }
                    d[word2[j]] = 1 + (d[word2[j]] ?? 0)
                }
                break
            }
            if j == len && word1.count > word2.count {
                return ""
            }
        }
        
        var q: [Character] = []
        for (x, y) in d {
            if y == 0 {
                q.append(x)
            }
        }
        
        var ans = ""
        while !q.isEmpty {
            let t = q.removeFirst()
            ans.append(t)
            if g[t] == nil {
                g[t] = []
            }
            for c in g[t]! {
                d[c] = (d[c] ?? 0) - 1
                if d[c] == 0 {
                    q.append(c)
                }
            }
        }
        return ans.count == g.count ? ans : ""
    }
}
// @lc code=end

func main() {
    var words: [String]
    var res: String
    
    /**
     输入：words = ["wrt","wrf","er","ett","rftt"]
     输出："wertf"
     */
    words = ["wrt","wrf","er","ett","rftt"]
    res = "wertf"
    assert(res == Solution().alienOrder(words))
    
    /**
     输入：words = ["z","x"]
     输出："zx"
     */
    words = ["z","x"]
    res = "zx"
    assert(res == Solution().alienOrder(words))
    
    /**
     输入：words = ["z","x","z"]
     输出：""
     解释：不存在合法字母顺序，因此返回 "" 。
     */
    words = ["z","x","z"]
    res = ""
    assert(res == Solution().alienOrder(words))
}

File: Cpp/剑指 Offer II/114.外星文字典.cpp
//
//  114.外星文字典.cpp
//  cpp
//
//  Created by kingsoft on 2022/5/31.
//

#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <queue>

using namespace std;

// @lc code=start
class Solution {
public:
    string alienOrder(vector<string>& words) {
        // 邻接矩阵
        unordered_map<char, unordered_set<char>> g;
        // 存每个字母的入度
        unordered_map<char, int> d;
        // 初始化每个字母的入度
        for (auto w: words)
            for (auto c: w)
                d[c] = 0;
        
        for (int i = 1; i < words.size(); ++i) {
            auto w1 = words[i - 1], w2 = words[i];
            int len = (int)min(w1.size(), w2.size());
            int j = 0;
            for (; j < len; ++j) {
                if (w1[j] == w2[j])
                    continue;
                if (!g.count(w1[j]) || !g[w1[j]].count(w2[j])) {
                    g[w1[j]].insert(w2[j]);
                    ++d[w2[j]];
                }
                // 找到第一个不同的字母（或没建边的点）后，直接退出
                break;    // 关键步
            }
            // 根据定义，此时w2的字典序小于w1的字典序，矛盾，说明无解
            if (j == len && w1.size() > w2.size())
                return "";
        }
        queue<char> q;
        for (auto [x, y] : d) {
            if (!y)
                q.push(x);
        }
        string res = "";
        while (!q.empty()) {
            auto t = q.front();
            q.pop();
            res += t;
            for (auto c: g[t]) {
                if (--d[c] == 0)
                    q.push(c);
            }
        }
        return res.size() == g.size() ? res : "";
    }
};
// @lc code=end

int main(int argc, const char * argv[]) {
    // insert code here...
    
    vector<string> words;
    string res;
    
    /**
     输入：words = ["wrt","wrf","er","ett","rftt"]
     输出："wertf"
     */
    words = {"wrt","wrf","er","ett","rftt"};
    res = "wertf";
    assert(res == Solution().alienOrder(words));
    
    /**
     输入：words = ["z","x"]
     输出："zx"
     */
    words = {"z","x"};
    res = "zx";
    assert(res == Solution().alienOrder(words));
    
    /**
     输入：words = ["z","x","z"]
     输出：""
     解释：不存在合法字母顺序，因此返回 "" 。
     */
    words = {"z","x","z"};
    res = "";
    assert(res == Solution().alienOrder(words));

    cout << "OK~" << endl;
    
    return 0;
}
